src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -86,12 +88,33 @@ public class TopicPartitionChannel {
   // This offset is updated when Snowflake has received offset from insertRows API
   // We will update this value after calling offsetToken API for this channel
   // We will only update it during start of the channel initialization
-  private long offsetPersistedInSnowflake = NO_OFFSET_TOKEN_REGISTERED_IN_SNOWFLAKE;
+  private AtomicLong offsetPersistedInSnowflake =
+      new AtomicLong(NO_OFFSET_TOKEN_REGISTERED_IN_SNOWFLAKE);
 
   // used to communicate to the streaming ingest's insertRows API
   // This is non final because we might decide to get the new instance of Channel
   private SnowflakeStreamingIngestChannel channel;
 
+  /**
+   * Offsets are reset in kafka when one of following cases arises in which we rely on source of
+   * truth (Which is Snowflake's committed offsetToken)
+   *
+   * <ol>
+   *   <li>If channel fails to fetch offsetToken in kafka, we reopen the channel and try to fetch
+   *       offset from Snowflake again
+   *   <li>If channel fails to ingest a buffer(Buffer containing rows/offsets), we reopen the
+   *       channel and try to fetch offset from Snowflake again
+   * </ol>
+   *
+   * <p>In both cases above, we ask Kafka to send back offsets, strictly from offset number after
+   * the offset present in Snowflake. i.e of Snowflake has offsetToken = x, we are asking Kafka to
+   * start sending offsets again from x + 1
+   *
+   * <p>This boolean is used to indicate that we reset offset in kafka and we will only buffer once
+   * we see the offset which is one more than an offset present in Snowflake.
+   */
+  private AtomicBoolean isOffsetResetInKafka = new AtomicBoolean(false);
```

Comment:
When we chatted offline about this we discussed whether this should be `AtomicBoolean` and if that's representative of a code smell because of potential threading issues that we may not be aware of. Do we need to use `AtomicBoolean` here or can we instead use `boolean`?

Reply:
Yes, per partition logic need not be thread safe but for sake of not mixing two things, I went ahead with keeping everything threadsafe. 

I will open up another PR for making all related changes.